enclave/core/sgx/thread.c

```diff
@@ -167,35 +167,109 @@ typedef struct _oe_mutex_impl
     /* The thread that has locked this mutex */
     oe_sgx_td_t* owner;
 
+    /* The type of mutex (supported types: normal and recursive) */
+    uint32_t type;
+
     /* Queue of waiting threads (front holds the mutex) */
     Queue queue;
 } oe_mutex_impl_t;
 
+typedef struct _oe_mutexattr_impl
+{
+    uint32_t type;
+} oe_mutexattr_impl_t;
+
 OE_STATIC_ASSERT(sizeof(oe_mutex_impl_t) <= sizeof(oe_mutex_t));
+OE_STATIC_ASSERT(sizeof(oe_mutexattr_impl_t) <= sizeof(oe_mutexattr_t));
 
-oe_result_t oe_mutex_init(oe_mutex_t* mutex)
+static oe_result_t _validate_mutexattr_type(uint32_t type)
 {
-    oe_mutex_impl_t* m = (oe_mutex_impl_t*)mutex;
+    oe_result_t result = OE_FAILURE;
+
+    switch (type)
+    {
+        case OE_MUTEX_NORMAL:
+        case OE_MUTEX_RECURSIVE:
+        case OE_MUTEX_ERRORCHECK:
+            break;
+        default:
+            OE_RAISE(OE_INVALID_PARAMETER);
+    };
+
+    result = OE_OK;
+
+done:
+    return result;
+}
+
+oe_result_t oe_mutexattr_init(oe_mutexattr_t* attr)
+{
+    oe_mutexattr_impl_t* attr_impl = (oe_mutexattr_impl_t*)attr;
+    oe_result_t result = OE_FAILURE;
+
+    if (!attr_impl)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    attr_impl->type = OE_MUTEX_DEFAULT;
+
+    result = OE_OK;
+
+done:
+    return result;
+}
+
+oe_result_t oe_mutexattr_settype(oe_mutexattr_t* attr, int type)
+{
+    oe_mutexattr_impl_t* attr_impl = (oe_mutexattr_impl_t*)attr;
+    oe_result_t result = OE_FAILURE;
+
+    if (!attr_impl)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    OE_CHECK(_validate_mutexattr_type((uint32_t)type));
+
+    attr_impl->type = (uint32_t)type;
+
+    result = OE_OK;
+
+done:
+    return result;
+}
+
+oe_result_t oe_mutex_init(oe_mutex_t* mutex, oe_mutexattr_t* attr)
+{
+    oe_mutex_impl_t* mutex_impl = (oe_mutex_impl_t*)mutex;
+    oe_mutexattr_impl_t* attr_impl = (oe_mutexattr_impl_t*)attr;
     oe_result_t result = OE_UNEXPECTED;
 
-    if (!m)
-        return OE_INVALID_PARAMETER;
+    if (!mutex_impl)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    memset(mutex_impl, 0, sizeof(oe_mutex_t));
+    mutex_impl->lock = OE_SPINLOCK_INITIALIZER;
 
-    memset(m, 0, sizeof(oe_mutex_t));
-    m->lock = OE_SPINLOCK_INITIALIZER;
+    if (!attr_impl)
+        mutex_impl->type = OE_MUTEX_DEFAULT;
+    else
+    {
+        OE_CHECK(_validate_mutexattr_type(attr_impl->type));
+
+        mutex_impl->type = attr_impl->type;
+    }
 
     result = OE_OK;
 
+done:
     return result;
 }
 
 /* Caller manages the spinlock */
 static int _mutex_lock(oe_mutex_impl_t* m, oe_sgx_td_t* self)
 {
     /* If this thread has already locked the mutex */
-    if (m->owner == self)
+    if (m->owner == self && m->type == OE_MUTEX_RECURSIVE)
```

Comment:
It would be good to detect recursive lock of a non-recursive mutex. In addition to logging it, should we also abort?